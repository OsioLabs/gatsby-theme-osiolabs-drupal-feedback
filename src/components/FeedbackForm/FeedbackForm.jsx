import React, {useState} from 'react';
import { Form, Message, Radio, TextArea } from 'semantic-ui-react';

const FeedbackForm = ({drupalOauthClient, tutorialId, tutorialUrl}) => {
  const [yesNo, setYesNo] = useState(false);
  const [feedback, setFeedback] = useState('');
  const [formState, setFormState] = useState({state: false, message: ''});

  // Get a CSRF token from Drupal.
  const getCSRFToken = async () => {
    const csrfTokenUrl = `${process.env.GATSBY_DRUPAL_API_ROOT}/session/token`;

    let csrfToken;
    try {
      const response = await fetch(csrfTokenUrl);
      if (response.ok) {
        csrfToken = await response.text();
      }
    } catch (e) {
      csrfToken = false;
    }

    return csrfToken;
  };

  // See if we can get an OAuth token. If so, we can authenticate the request.
  const getOauthToken = async () => {
    let oauthToken;

    try {
      oauthToken = await drupalOauthClient.isLoggedIn();
    } catch (e) {
      oauthToken = false;
    }

    return oauthToken;
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    setFormState({state: 'loading'});

    const feedbackPostUrl = `${process.env.GATSBY_DRUPAL_API_ROOT}/webform_rest/submit?_format=json`;
    const csrfToken = await getCSRFToken();

    // This isn't going to work without a CSRF token to bail early if we don't
    // have one.
    if (!csrfToken) {
      setFormState({state: 'error', message: 'Invalid CSRF token.'});
      return;
    }

    const oauthToken = await getOauthToken();

    const headers = new Headers({
      Accept: 'application/json',
      'Content-Type': 'application/json',
      'X-CSRF-Token': csrfToken,
      'X-Consumer-ID': `${process.env.GATSBY_DRUPAL_API_ID}`,
    });

    if (oauthToken) {
      headers.set('Authorization', `${oauthToken.token_type} ${oauthToken.access_token}`);
    }

    const payload = {
      webform_id: 'tutorial_feedback',
      was_this_helpful: yesNo,
      feedback: feedback,
      tutorial_url: `${process.env.GATSBY_ROOT_URL}${tutorialUrl}`,
      tutorial_id: tutorialId
    };

    const options = {
      method: 'POST',
      headers,
      body: JSON.stringify(payload),
    };

    try {
      const response = await fetch(feedbackPostUrl, options);
      if (response.ok) {
        // Success.
        const data = await response.json();

        if (typeof data.error !== 'undefined') {
          // Error processing the request.
          setFormState({state: 'error', message: data.error.message});
        } else {
          // Clear the form, and display success message.
          setYesNo(false);
          setFeedback('');
          setFormState({state: 'success'});
        }
      } else {
        // Error connecting to Drupal.
        setFormState({state: 'error', message: `${response.status}: ${response.statusText}`});
      }
    } catch (e) {
      setFormState({state: 'error', message: e.message});
    }
  };

  return (
    <Form
      onSubmit={handleSubmit}
      error={formState.state === 'error'}
      success={formState.state === 'success'}
      loading={formState.state === 'loading'}
      className="feedbackform"
    >
      {formState.state === 'success' &&
        <Message success content="Thanks for your feedback!"/>
      }
      {formState.state === 'error' &&
      <Message error header="Oh no, something went wrong"
               content={formState.message}/>
      }
      <Form.Group className="feedbackform--yesno">
        <div className="feedbackform--yesno-label">Was this helpful?</div>
        <Form.Field>
          <Radio
            label='Yes'
            name='yesNo'
            id='yesNo-yes'
            value='Yes'
            checked={yesNo === 'Yes'}
            onChange={event => setYesNo('Yes')}
          />
        </Form.Field>
        <Form.Field>
          <Radio
            label='No'
            name='yesNo'
            id='yesNo-no'
            value='No'
            checked={yesNo === 'No'}
            onChange={event => setYesNo('No')}
          />
        </Form.Field>
      </Form.Group>
      {yesNo &&
        <Form.Group className="feedbackform--hidden">
          <label htmlFor="feedback">Additional feedback:</label>
          <TextArea
            name="feedback"
            value={feedback}
            onChange={event => setFeedback(event.target.value)}
            placeholder='Tell us more ...'
          />
          <Form.Button content='Share feedback' secondary />
        </Form.Group>
      }
    </Form>
  );
};

export default FeedbackForm;
