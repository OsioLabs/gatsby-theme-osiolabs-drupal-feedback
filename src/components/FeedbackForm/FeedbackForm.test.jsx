import React from 'react';
import { act, cleanup, render, fireEvent, waitForElement } from '@testing-library/react';
import FeedbackForm from './FeedbackForm';

import DrupalOauth from 'gatsby-theme-osiolabs-drupal/src/components/drupal-oauth/DrupalOauth';
jest.mock('gatsby-theme-osiolabs-drupal/src/components/drupal-oauth/DrupalOauth');
const drupalOauthClient = new DrupalOauth();

afterEach(cleanup);

beforeEach(() => {
  fetch.resetMocks();
  drupalOauthClient.isLoggedIn.mockClear();
});

/**
 * Helper to fill out the feedback form.
 */
const doFillForm = async (getByText, getByLabelText, getByPlaceholderText) => {
  const yesNoField = getByLabelText('Yes');
  fireEvent.click(yesNoField);
  const feedbackField = await waitForElement(() =>
    getByPlaceholderText('Tell us more ...')
  );
  fireEvent.change(feedbackField, { target: { value: 'It does not suck.' } });
  const submitButton = getByText('Share feedback');
  fireEvent.click(submitButton);
};

describe('Component: <FeedbackForm />', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <FeedbackForm
        drupalOauthClient={drupalOauthClient}
        tutorialId="42"
        tutorialUrl="http://example.com/tutorial"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays a success message', async () => {
    fetch.mockResponse(JSON.stringify({ sid: '12345' }));

    const { container, getByText, getByLabelText, getByPlaceholderText } = render(
      <FeedbackForm
        drupalOauthClient={drupalOauthClient}
        tutorialId="42"
        tutorialUrl="http://example.com/tutorial"
      />
    );

    // Submit the form.
    await doFillForm(getByText, getByLabelText, getByPlaceholderText);

    expect(container.firstChild).toHaveClass('loading');

    const success = await waitForElement(() =>
      getByText(/Thanks for your feedback/)
    );

    expect(success).toBeTruthy();
    // Once to see if the user is logged and to try and get an oauth token.
    expect(drupalOauthClient.isLoggedIn).toHaveBeenCalledTimes(1);
    // Once to get a CSRF token, and once to submit the data to the API.
    expect(fetch).toHaveBeenCalledTimes(2);
  });

  it('displays error messages returned from the server', async () => {
    fetch
      .once("CSRF_TOKEN")
      .once(JSON.stringify({ error: {message: 'Mock error from server'} }));

    const { container, getByText, getByLabelText, getByPlaceholderText } = render(
      <FeedbackForm
        drupalOauthClient={drupalOauthClient}
        tutorialId="42"
        tutorialUrl="http://example.com/tutorial"
      />
    );

    // Submit the form.
    await doFillForm(getByText, getByLabelText, getByPlaceholderText);

    expect(container.firstChild).toHaveClass('loading');

    const error = await waitForElement(() =>
      getByText(/Mock error from server/)
    );

    expect(error).toBeTruthy();
    // Once to see if the user is logged and to try and get an oauth token.
    expect(drupalOauthClient.isLoggedIn).toHaveBeenCalledTimes(1);
    // Once to get a CSRF token, and once to submit the data to the API.
    expect(fetch).toHaveBeenCalledTimes(2);
  });

  it("displays an error message when API can't be reached", async () => {
    fetch
      .once('CSRF_TOKEN')
      .mockReject(new Error('mock api unreachable'));

    const { container, getByText, getByLabelText, getByPlaceholderText } = render(
      <FeedbackForm
        drupalOauthClient={drupalOauthClient}
        tutorialId="42"
        tutorialUrl="http://example.com/tutorial"
      />
    );

    // Submit the form.
    await doFillForm(getByText, getByLabelText, getByPlaceholderText);

    const error = await waitForElement(() =>
      getByText(/mock api unreachable/)
    );

    expect(error).toBeTruthy();
    expect(fetch).toHaveBeenCalledTimes(2);

    // Try it again with a BAD CSRF token.
    fetch.resetMocks();
    fetch.mockReject(new Error('no csrf token'));

    // Submit the form.
    await doFillForm(getByText, getByLabelText, getByPlaceholderText);

    const csrfError = await waitForElement(() =>
      getByText(/Invalid CSRF token/)
    );

    expect(csrfError).toBeTruthy();
    // We bail early and don't try and submit to the API if we don't have a
    // valid CSRF token.
    expect(fetch).toHaveBeenCalledTimes(1);
  });
});
