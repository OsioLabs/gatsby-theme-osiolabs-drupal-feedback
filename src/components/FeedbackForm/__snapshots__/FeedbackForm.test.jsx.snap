// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <FeedbackForm /> renders correctly 1`] = `
<DocumentFragment>
  <form
    class="ui form feedbackform"
  >
    <div
      class="fields feedbackform--yesno"
    >
      <div
        class="feedbackform--yesno-label"
      >
        Was this helpful?
      </div>
      <div
        class="field"
      >
        <div
          class="ui radio checkbox"
        >
          <input
            class="hidden"
            id="yesNo-yes"
            name="yesNo"
            readonly=""
            tabindex="0"
            type="radio"
            value="Yes"
          />
          <label
            for="yesNo-yes"
          >
            Yes
          </label>
        </div>
      </div>
      <div
        class="field"
      >
        <div
          class="ui radio checkbox"
        >
          <input
            class="hidden"
            id="yesNo-no"
            name="yesNo"
            readonly=""
            tabindex="0"
            type="radio"
            value="No"
          />
          <label
            for="yesNo-no"
          >
            No
          </label>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
